synse/http.go

```diff
@@ -112,28 +117,222 @@ func (c *httpClient) Config() (*Config, error) {
+// ReadDevice returns data from a specific device. It is the same as Read()
+// where the label matches the device id tag specified in ReadOptions.
+func (c *httpClient) ReadDevice(id string, opts scheme.ReadOptions) (*[]scheme.Read, error) {
+	out := new([]scheme.Read)
+	err := c.getVersionedQueryParams(makeURI(readURI, id), opts, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/read` endpoint")
+	}
+
+	return out, nil
+}
+
+// ReadCache returns stream reading data from the registered plugins.
+func (c *httpClient) ReadCache(opts scheme.ReadCacheOptions) (*[]scheme.Read, error) {
+	var out []scheme.Read
+	errScheme := new(scheme.Error)
+
+	client, err := c.setVersioned()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to set a versioned host while performing a GET request")
+	}
+
+	resp, err := client.R().SetDoNotParseResponse(true).SetQueryParams(structToMapString(opts)).SetError(errScheme).Get(readcacheURI)
+	err = check(err, errScheme)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/readcache` endpoint")
+	}
+
+	dec := json.NewDecoder(resp.RawBody())
+	for dec.More() {
+		var read scheme.Read
+		err := dec.Decode(&read)
+		if err != nil {
+			return nil, errors.Wrap(err, "failed to decode a JSON response into an appropriate struct")
+		}
+
+		out = append(out, read)
+	}
+
+	return &out, nil
+}
+
+// Write writes data to a device, in an asynchronous manner.
+func (c *httpClient) Write(id string, opts []scheme.WriteData) (*[]scheme.Write, error) {
+	out := new([]scheme.Write)
+	err := c.postVersioned(makeURI(writeURI, id), opts, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/write` endpoint")
+	}
+
+	return out, nil
+}
+
+// WriteWait writes data to a device, waiting for the write to complete.
+func (c *httpClient) WriteWait(id string, opts []scheme.WriteData) (*[]scheme.Transaction, error) {
+	out := new([]scheme.Transaction)
+	err := c.postVersioned(makeURI(writeWaitURI, id), opts, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/write/wait` endpoint")
+	}
+
+	return out, nil
+}
+
+// Transactions returns the sorted list of all cached transaction IDs.
+func (c *httpClient) Transactions() (*[]string, error) {
+	out := new([]string)
+	err := c.getVersioned(transactionURI, out)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to request `/transaction` endpoint")
+	}
+
+	return out, nil
+}
+
+// Transaction returns the state and status of a write transaction.
+func (c *httpClient) Transaction(id string) (*scheme.Transaction, error) {
+	out := new(scheme.Transaction)
+	err := c.getVersioned(makeURI(transactionURI, id), out)
+	if err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to request `/transaction/%v` endpoint", id))
+	}
+
+	return out, nil
+}
+
+// Metrics returns the application-based metrics.
+func (c *httpClient) Metrics() (*scheme.Metrics, error) {
+	return nil, errors.New("not yet implemented")
+}
```

Comment:
I think this can be removed. The metrics endpoint isn't really an API endpoint, its meant for metrics collection via something like prometheus which will export the data in its own format that isn't useful to anything other than prometheus